File: Top_100_Liked_Questions/394._Decode_String/solution.js
/**
 * @param {string} s
 * @return {string}
 */
var decodeString = function(s) {
    let decode = function(s, rep) {
        let stack = 0;
        let i = 0;
        let found = false;
        let num_start = 0;
        let par_start = 0;
        let num = 0;
        
        while (i < s.length) {
            if (found) {
                if (s[i] === '[') {
                    stack++;
                    if (par_start === 0) {
                        par_start = i;
                    }
                    if (num === 0) {
                        num = parseInt(s.slice(num_start, i));
                    }
                } else if (s[i] === ']') {
                    stack--;
                    if (stack === 0) {
                        let sub_s = decode(s.slice(par_start+1, i), num);
                        s = s.slice(0, num_start) + sub_s + s.slice(i+1);
                        i += sub_s.length - (i - num_start + 1);
                        found = false;
                        num_start = 0;
                        par_start = 0;
                        num = 0;
                    }
                }
            } else {
                if (!isNaN(parseInt(s[i]))) {
                    found = true;
                    num_start = i;
                }
            }
            
            i++;
        }
        
        return s.repeat(rep);
    }
    
    return decode(s, 1);
};

File: Top_100_Liked_Questions/394._Decode_String/solution.py
class Solution:
    def decodeString(self, s: str) -> str:
        def decode(s, rep):
            # find outer bracket, i.e closing bracket that pops the stack empty, and parse the number of repetitions and its inner string
            # call recursively with the args and swap the substring portion with the returned substring
            # continue until the end of the string and return the final string times repetition param
            stack = 0
            i = 0
            found = False
            par_start = num_start = num = 0
            while i < len(s):
                if found:
                    if s[i] == '[':
                        if par_start == 0:
                            par_start = i
                        if num == 0:
                            num = int(s[num_start:i])
                        stack += 1
                    elif s[i] == ']':
                        stack -= 1
                        if stack == 0:
                            sub_s = decode(s[par_start+1:i], num)
                            s = s[:num_start] + sub_s + s[i+1:]
                            i += len(sub_s) - (i - num_start + 1)
                            found = False
                            par_start = 0
                            num = 0
                else:
                    if s[i].isdigit():
                        found = True
                        num_start = i
                i += 1
                
            return s * rep
        
        return decode(s, 1)